shopping-cart/shopping-cart-scala/shopping-cart/src/main/scala/com/example/shoppingcart/impl/ShoppingCart.scala

```diff
@@ -0,0 +1,250 @@
+    override def aggregateTag: AggregateEventTagger[Event] = Event.Tag
+  }
+
+  object Event {
+    val Tag: AggregateEventShards[Event] = AggregateEventTag.sharded[Event](numShards = 10)
+  }
+
+  final case class ItemAdded(itemId: String, quantity: Int) extends Event
+
+  final case class ItemRemoved(itemId: String) extends Event
+
+  final case class ItemQuantityAdjusted(itemId: String, newQuantity: Int) extends Event
+
+  final case class CartCheckedOut(eventTime: Instant) extends Event
+
+  // Events get stored and loaded from the database, hence a JSON format
+  //  needs to be declared so that they can be serialized and deserialized.
+  implicit val itemAddedFormat: Format[ItemAdded]                       = Json.format
+  implicit val itemRemovedFormat: Format[ItemRemoved]                   = Json.format
+  implicit val itemQuantityAdjustedFormat: Format[ItemQuantityAdjusted] = Json.format
+  implicit val cartCheckedOutFormat: Format[CartCheckedOut]             = Json.format
+
+  // SHOPPING CART REPLIES
+  final case class Summary(items: Map[String, Int], checkedOut: Boolean)
+
+  sealed trait Confirmation
+
+  final case class Accepted(summary: Summary) extends Confirmation
+
+  final case class Rejected(reason: String) extends Confirmation
+
+  implicit val summaryFormat: Format[Summary]               = Json.format
+  implicit val confirmationAcceptedFormat: Format[Accepted] = Json.format
+  implicit val confirmationRejectedFormat: Format[Rejected] = Json.format
+  implicit val confirmationFormat: Format[Confirmation] = new Format[Confirmation] {
+    override def reads(json: JsValue): JsResult[Confirmation] = {
+      if ((json \ "reason").isDefined)
+        Json.fromJson[Rejected](json)
+      else
+        Json.fromJson[Accepted](json)
+    }
+
+    override def writes(o: Confirmation): JsValue = {
+      o match {
+        case acc: Accepted => Json.toJson(acc)
+        case rej: Rejected => Json.toJson(rej)
+      }
+    }
+  }
+
+  val empty = ShoppingCart(Map.empty, checkedOut = false)
+
+  val typeKey: EntityTypeKey[Command] = EntityTypeKey[Command]("ShoppingCart")
+
+  // We can then access the entity behavior in our test tests, without the need to tag
+  // or retain events.
+  def behavior(persistenceId: PersistenceId): EventSourcedBehavior[Command, Event, ShoppingCart] = {
+    EventSourcedBehavior
+      .withEnforcedReplies[Command, Event, ShoppingCart](
+        persistenceId = persistenceId,
+        emptyState = ShoppingCart.empty,
+        commandHandler = (cart, cmd) => cart.applyCommand(cmd),
+        eventHandler = (cart, evt) => cart.applyEvent(evt)
+      )
+  }
+
+  def behavior(entityContext: EntityContext[Command]): Behavior[Command] =
+    behavior(PersistenceId(entityContext.entityTypeKey.name, entityContext.entityId))
+      .withTagger(AkkaTaggerAdapter.fromLagom(entityContext, Event.Tag))
+      .withRetention(RetentionCriteria.snapshotEvery(numberOfEvents = 100, keepNSnapshots = 2))
```

Comment:
This is introducing yet another convention. In the Akka style guide for scala code, the convention is to name these factory methods `apply`. 

We (@renatocaval and I) have been lobbying in favor of using for scala code the same sonvention that Akka style guide promotes for java code: `def create(...):Behavior[_]`.